lib/fs.js

```diff
@@ -1400,6 +1427,12 @@ function mkdirSync(path, options) {
 function readdir(path, options, callback) {
   callback = makeCallback(typeof options === 'function' ? options : callback);
   options = getOptions(options);
+
+  if (options.recursive) {
+    callback(null, readdirSyncRecursive(path, options));
```

Comment:
Calling a recursive synchronous function from an asynchronous API seems like something that is almost certainly going to cause problems.

Reply:
I've bounced between solutions here. @mcollina expressed concerns previously about my recursive asynchronous solution could result in too many promises being created. We also agreed that shipping _some_ solution is better than nothing and that once its landed we can work on improving it

Reply:
> expressed concerns previously about my recursive asynchronous solution could result in too many promises being created

I don't know if you've benchmarked or not, but this seems likely to be worse. We added `opendir()` and friends because a single `readdir()` could be too costly in a large directory. This has the potential to be much, much slower while also being synchronous, while also misleading users about the real nature of the API.